modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpRequestBodyStreamTests.java

```diff
@@ -128,57 +134,112 @@ public void testReadFromChannel() {
-        stream.next();
+            eventLoop.submit(() -> channel.runPendingTasks()).await();
+            assertThat(headers.get(), hasEntry("header1", "value1"));
+            assertThat(headers.get(), hasEntry("header2", "value2"));
+            assertThat(headers.get(), hasEntry("header3", "value3"));
 
-        thread = new Thread(() -> channel.runPendingTasks());
-        thread.start();
-        thread.join();
+            assertTrue("should receive last content", gotLast.get());
 
-        assertThat(headers.get(), hasEntry("header1", "value1"));
-        assertThat(headers.get(), hasEntry("header2", "value2"));
-        assertThat(headers.get(), hasEntry("header3", "value3"));
+            headers.set(new HashMap<>());
 
-        assertTrue("should receive last content", gotLast.get());
+            stream.close();
+
+            assertThat(headers.get(), hasEntry("header1", "value1"));
+            assertThat(headers.get(), hasEntry("header2", "value2"));
+            assertThat(headers.get(), hasEntry("header3", "value3"));
+        } finally {
+            eventLoop.shutdownGracefully(0, 0, TimeUnit.SECONDS);
+        }
+    }
 
-        headers.set(new HashMap<>());
+    public void testStreamNextActivityTracker() {
+        var t0 = activityTracker.get();
+        var N = between(1, 10);
+        for (int i = 0; i < N; i++) {
+            channel.writeInbound(randomContent(1024));
+            stream.next();
+            channel.runPendingTasks();
+        }
+        var t1 = activityTracker.get();
+        assertEquals("stream#next() must trigger activity tracker: N*step=" + N + "*2=" + N * 2L + " times", t1, t0 + N * 2L);
+    }
 
-        stream.close();
+    // ensure that we catch all exceptions and throw them into channel pipeline
+    public void testCatchExceptions() {
+        var gotExceptions = new CountDownLatch(3); // number of tests below
+
+        channel.pipeline().addLast(new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+                gotExceptions.countDown();
+            }
+        });
+
+        // catch exception for not buffered chunk, will be thrown on channel.fireChannelRead()
+        stream.setHandler((a, b) -> { throw new RuntimeException(); });
+        stream.next();
+        channel.runPendingTasks();
+        channel.writeInbound(randomContent(1));
+
+        // catch exception for buffered chunk, will be thrown from eventLoop.submit()
+        channel.writeInbound(randomContent(1));
+        stream.next();
+        channel.runPendingTasks();
+
+        // should catch OOM exceptions too, see DieWithDignity
+        // swallowing exceptions can result in dangling streams, hanging channels, and delayed shutdowns
+        stream.setHandler((a, b) -> { throw new OutOfMemoryError(); });
+        channel.writeInbound(randomContent(1));
+        stream.next();
+        channel.runPendingTasks();
```

Comment:
This seems to test handling `Error`s for a buffered read, do we need to add a test that we do the same for a not-buffered chunk (I think we broadened the catch to Throwable from Exception in this case)?

Reply:
Oh no we didn't, that's a separate code path (channelRead). Never mind.